evap/staff/forms.py

```diff
@@ -387,6 +387,11 @@ def send(self, request, courses):
 
 
 class QuestionnaireForm(forms.ModelForm):
+    def save(self, *args, **kwargs):
+        if self.cleaned_data['type'] != self.instance.type:
+            self.cleaned_data['order'] = Questionnaire.objects.filter(type=self.cleaned_data['type']).aggregate(Max('order'))['order__max'] + 1
```

Comment:
does the condition work correctly when creating a new questionnaire?

Reply:
When you're creating a new questionnaire, this doesn't make a difference since the order will be overriden in the questionnaire_create staff view.

This does not work when changing an existing questionnaire - it was still on my todo list. I'm not sure about whether I want to do what this code implies though. When you edit a questionnaire, you can edit both, the type and the order value. Now, if you have a questionnaire that's on the top of the list and you go to the edit page, you will have the values type top and order 0 in the input fields. If you just change the type and click submit, wouldn't it be strange if the order changed from 0 to to highest value?
Or even more strange, if it was the last and you wanted it to be the first, so you changed the type from top to bottom and the order from whatever to 0. You submit the form and the questionnaire is back on the bottom with a high order value.

From this perspective, I would not touch the order. It's not a big problem if we have duplicate order values in one type and reordering them with the drag-and-drop interface will fix up every messy value.

Also strange: When a questionnaire is in use, you can not edit it's order value on the edit page but you can change it with the drag-and-drop interface.

@janno42 what's your opinion?

Reply:
you could check whether the instance has a pk and if not, assume this is a newly created questionnaire, and then do the number assignment here instead of in the view. i think that would be nicer.

for existing questionnaires, having the order in the form is weird i think because the user has little information what this means. even if they know how the sorting works, the only options that make any sense are zero (meaning "either above or below the previous top questionnaire" or "some large number" meaning at the bottom.

the simplest solution would probably be to remove the order field from the form, always put it to the bottom, and maybe output a message to the user after saving that they need to check the ordering.

Reply:
What's the order field you're talking about? Did you add that to the form? Nobody will ever use such a field in the form. Ordering only happens via staff and drop on the overview page. Users will expect that new questionnaires will be added on the bottom and when changing the "type" moving it to the bottom (preferably with the message) is also okay.

Reply:
Ahh now that makes sense. Yes, I added it - by accident, when I renamed the "index" field to "order". The fields that are excluded in the form are only referenced by name as a string, so nothing broke and I didn't notice. Alright, so it'll just be excluded. Never mind then.